lxd/storage/drivers/driver_dir.go

```diff
@@ -0,0 +1,293 @@
+	return false
+}
+
+// HasVolume indicates whether a specific volume exists on the storage pool.
+func (d *dir) HasVolume(volType VolumeType, volName string) bool {
+	if shared.PathExists(GetVolumeMountPoint(d.name, volType, volName)) {
+		return true
+	}
+
+	return false
+}
+
+func (d *dir) Create() error {
+	// WARNING: The Create() function cannot rely on any of the struct attributes being set.
+
+	// Set default source if missing.
+	if d.config["source"] == "" {
+		d.config["source"] = GetPoolMountPoint(d.name)
+	}
+
+	if !shared.PathExists(d.config["source"]) {
+		return fmt.Errorf("Source path '%s' doesn't exist", d.config["source"])
+	}
+
+	// Check that the path is currently empty.
+	isEmpty, err := shared.PathIsEmpty(d.config["source"])
+	if err != nil {
+		return err
+	}
+
+	if !isEmpty {
+		return fmt.Errorf("Source path '%s' isn't empty", d.config["source"])
+	}
+
+	return nil
+}
+
+func (d *dir) Delete(op *operations.Operation) error {
+	// On delete, wipe everything in the directory.
+	err := wipeDirectory(GetPoolMountPoint(d.name))
+	if err != nil {
+		return err
+	}
+
+	// Unmount the path.
+	_, err = d.Unmount()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (d *dir) Mount() (bool, error) {
+	path := GetPoolMountPoint(d.name)
+
+	// Check if we're dealing with an external mount.
+	if d.config["source"] == path {
+		return false, nil
+	}
+
+	// Check if already mounted.
+	if sameMount(d.config["source"], path) {
+		return false, nil
+	}
+
+	// Setup the bind-mount.
+	err := tryMount(d.config["source"], path, "none", unix.MS_BIND, "")
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (d *dir) Unmount() (bool, error) {
+	path := GetPoolMountPoint(d.name)
+
+	// Check if we're dealing with an external mount.
+	if d.config["source"] == path {
+		return false, nil
+	}
+
+	// Unmount until nothing is left mounted.
+	return forceUnmount(path)
+}
+
+func (d *dir) GetResources() (*api.ResourcesStoragePool, error) {
+	// Use the generic VFS resources.
+	return vfsResources(GetPoolMountPoint(d.name))
+}
+
+// DeleteVolume deletes a volume of the storage device. If the volume supplied is a snapshot and
+// it is the last one being removed for the parent volume then the snapshots directory for that
+// parent volume is removed too.
+func (d *dir) DeleteVolume(volType VolumeType, volName string, op *operations.Operation) error {
+	volPath := GetVolumeMountPoint(d.name, volType, volName)
+
+	if !shared.PathExists(volPath) {
+		return nil
+	}
+
+	volID, err := d.getVolID(volType, volName)
+	if err != nil {
+		return err
+	}
+
+	err = d.deleteQuota(volPath, volID)
+	if err != nil {
+		return err
+	}
+
+	err = os.RemoveAll(volPath)
+	if err != nil {
+		return err
+	}
+
+	// If the volume being removed is a snapshot, check if it is the last one being removed for
+	// the volume parent and if so then remove the parent volume's snaphots directory.
+	if shared.IsSnapshot(volName) {
```

Comment:
I don't like the idea of us mixing normal volumes and snapshots together. It's something we've done everywhere in the past and that's come to bite us since. Things are also going to get weird with zfs, lvm and ceph where there is a strong distinction between volumes and snapshots.

Also be careful about quotas on dir backend, we want all snapshots to share the quota group of their volume, so we need to pass the volume name, not the snapshot name to getVoldID.

I think it'd be best if we had separate functions to create/delete/rename/restore snapshots.

We'll also want to make sure that deleting a volume which still has snapshots fails. This should never happen as the backend should do the right thing and delete the snapshots first, but we've seen weird inconsistencies before.

Reply:
Thinking about this again, it may actually be fine to use the same functions as we have distinct volume types now.

We however shouldn't be using shared.IsSnapshot as we can and should distinguish based on voltype.

Reply:
@stgraber there isn't a vol type for snapshot, so I'm not clear how these would help indicate a snapshot?

Reply:
Okay, so clearly my memory was failing me, I thought we had distinct volume types for regular vs snapshots, we don't, so we obviously can't rely on them. So I think that convinced me to go back to separate functions then, even if they end up internally sharing a lot of logic through shared private functions.

CreateVolumeSnapshot (take a snapshot)
DeleteVolumeSnapshot (delete a snapshot)
RenameVolumeSnapshot (rename a snapshot)

Reply:
@stgraber OK, I have added those functions to the interface and modified this function to not only delete the actual parent volume (and not the snapshots) as well as checking all the snapshots have been removed first.